### PR TITLE
Events Component

### PR DESCRIPTION
**Pull request by bafonins**
**Summary:**

This PR adds the `<Events />` component that will be used in the Entity Data pages.
References:  https://github.com/TheThingsNetwork/lorawan-stack/issues/28
Closes: https://github.com/TheThingsNetwork/lorawan-stack/issues/553

empty:
Screenshot 2019-04-25 at 15 39 07

regular:
Screenshot 2019-04-26 at 15 41 14

expanded:
Screenshot 2019-04-26 at 15 38 56

**Changes:**

- Changes to the `<Event />` component to support expanding view
- Add the `<Events />` component

**Notes for Reviewers:**

@kschiffer:
1. I wasnt able to virtualize the events list in case. There is too much of dynamic height calculation and the implementation I have doesnt seem to be robust, hence I decided to follow the V2 traffic approach.
2. The component doesnt doesnt limit/truncate the number of events on its own. This is mainly because of the pause functionality. I can imagine we want to control that from the outside.
3. Do we want to inform the user that events are being truncated?
4. Im using `React.PureComponent` in the event types components to avoid redundant rerenders.

@htdvisser :
1. Please check the UI. 
2. See pkg/webui/components/event/types/message/format-message.js to verify that we extract the right entries from the event `data` field for each event type.
3. Can we assume that a key of the form `{event.time}-{event.name}` can uniquely identify an event, since we dont consider historical events?

### CODE REVIEW

**Comment by htdvisser** [comment]
1. The UI looks good in general. Maybe it's better to use a different icon for connect/disconnect, the lightning bolt already indicates joins in v2. Also it's not clear enough which events have details and which don't
2. The "details" JSON viewer should contain the entire `data` object of an event and not just some selected fields (you can only leave out the `@type` field). In the future we will extract specific fields to show in a more friendly form, some even in the table rows, like we have in the v2 console.
3. It's not likely that in the same events view multiple events of the same type happen at the exact same microsecond, but it's not impossible. We could however add unique IDs to the Event type if you want.

File: pkg/webui/components/icon/index.js [hunk 1]
```diff
@@ -38,7 +38,7 @@ const hardcoded = {
   general_settings: 'settings',
   location: 'place',
   user: 'person',
-  event: 'offline_bolt',
+  event: 'wifi',
```
Comment (kschiffer) [inline pkg/webui/components/icon/index.js]: I think we still need to work on the event icons and likely rely on another icon set than material icons. Albeit good enough for now.

File: pkg/webui/components/list/index.js [hunk 2]
```diff
@@ -115,6 +139,7 @@ List.defaultProps = {
   bordered: false,
   emptyMessage: sharedMessages.noMatch,
   emptyMessageValues: {},
+  header: null,
```
Comment (kschiffer) [inline pkg/webui/components/list/index.js]: The `renderItem` function you have in the story should be the default prop here.

File: pkg/webui/console/views/application-data/index.js [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import React from 'react'
+import { Container, Col, Row } from 'react-grid-system'
+
+import Events from '../../../components/events'
+
+const events = [
```
Comment (kschiffer) [inline pkg/webui/console/views/application-data/index.js]: What's the purpose of this?

File: pkg/webui/console/views/application-data/index.js [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+  { name: 'gs.gateway.disconnect', time: '2019-04-15T09:20:39.679935Z', identifiers: [{ gateway_ids: { gateway_id: 'admin-gtw', eui: '0102030405060708' }}], correlation_ids: [ 'gs:conn:01D8G4RSYNH95CEQ0CY3Y6ESH8', 'rpc:/ttn.lorawan.v3.GtwGs/LinkGateway:01D8G4RSVNGBNQMGZ3Q1QJQFKK' ], origin: 'htdvissermbp' },
+]
+
+export default class Data extends React.Component {
```
Comment (kschiffer) [inline pkg/webui/console/views/application-data/index.js]: I believe this whole file is not supposed to be here, or at least not in this stage?

File: pkg/webui/components/events/events.styl [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+    &-id
+      flex-basis: 20%
+      min-width: 7rem
+      margin-left: 2.8rem
+      margin-right: $cs.s
+
+  &-actions
+    position: absolute
+    right: 2rem
```
Comment (kschiffer) [inline pkg/webui/components/events/events.styl]: Should rather attach right to the edge (as in `right: 0`). See also my other comment about handling the positioning here with flexbox.

File: pkg/webui/components/events/index.js [hunk 6]
```diff
@@ -13,17 +13,133 @@
...
+  const pauseIcon = paused ? 'play_arrow' : 'pause'
+
+  return (
+    <div className={style.header}>
+      <div className={style.headerColumns}>
+        <Message className={style.headerColumnsTime} content={sharedMessages.time} />
+        <Message className={style.headerColumnsId} content={sharedMessages.entityId} />
+        <Message content={sharedMessages.data} />
+      </div>
+      <div className={style.headerActions}>
```
[the remark this answers is not in the document]
Comment (bafonins) [reply pkg/webui/components/events/index.js]: No, the reason to have `position: absolute` here is to align entries from the list with the entries in the header.

[the remark this answers is not in the document]
Comment (kschiffer) [reply pkg/webui/components/events/index.js]: Ok, can you still consider this then: https://github.com/TheThingsNetwork/lorawan-stack/pull/570#discussion_r279206495

File: pkg/webui/components/events/events.styl [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+
+.footer
+  display: flex
+  justify-content: center
+  align-items: center
+  font-size: $fs.s
+
```
Comment (kschiffer) [inline pkg/webui/components/events/events.styl]:
```suggestion
color: $tc-subtle-gray
```

File: pkg/webui/components/list/index.js (hunk 2 above)
Comment (kschiffer) [reply pkg/webui/components/list/index.js]: I meant this one:
```js
  renderItem (item, index) {
    return (
      <List.Item key={index}>
        {item}
      </List.Item>
    )
  }
```

File: pkg/webui/components/events/events.styl (hunk 5 above)
Comment (kschiffer) [reply pkg/webui/components/events/events.styl]:
```suggestion
    right: 0
```

**Comment by bafonins** [comment]
> 1. Not sure if I get it. What exactly were you trying to achieve?

Tried to virtualize list entries. tl;dr it didnt work.

>In storybook, I see a lot of these errors:
`[React Intl] Cannot format message: "event:gs.down.send", using message id as fallback.`

Will create a separate issue for that and check if there is a quick fix for that

**Pull request merged**